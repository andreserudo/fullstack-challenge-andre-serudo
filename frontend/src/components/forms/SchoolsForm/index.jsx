import React, { useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import SchoolFormWrapper from './styles';
import { createSchool } from '../../../services/schools';
import ApplicationContext from '../../../context/ApplicationContext';

const defaultSchoolForm = {
  nome: '',
  responsavel: '',
};

const statusFormMessages = {
  success: 'Cadastro realizado com sucesso!',
  error: 'Não foi possível realizar o cadastro.',
};

function SchoolsForm({ closeForm }) {
  const { getAllSchools } = useContext(ApplicationContext);
  const [schoolValues, setSchoolValues] = useState(defaultSchoolForm);
  const [allowToSave, setAllowToSave] = useState(true);
  const [schoolCreated, setSchoolCreated] = useState(false);
  const [messageForSchoolSubmitted, setMessageForSchoolSubmitted] = useState('');

  const handleChange = (e, name) => {
    const { value } = e.target;
    setSchoolValues({ ...schoolValues, [name]: value });
  };

  const handleSubbmit = async (e) => {
    const { nome, responsavel } = schoolValues;
    e.preventDefault();

    const school = await createSchool({ nome, responsavel });
    setSchoolCreated(!schoolCreated);

    if (!school) {
      setMessageForSchoolSubmitted(statusFormMessages.error);
      return;
    }

    await getAllSchools();

    setSchoolValues(defaultSchoolForm);
  };

  const checkIfCanAllowSave = () => {
    const { nome, responsavel } = schoolValues;

    if ((nome !== '') && (responsavel !== '')) {
      setAllowToSave(false);
    } else {
      setAllowToSave(true);
    }
  };

  useEffect(() => {
    checkIfCanAllowSave();
  }, [schoolValues]);

  return (
    <SchoolFormWrapper>
      <form onSubmit={(e) => handleSubbmit(e)}>
        <label htmlFor="nome">
          Nome da Escola:
          <input
            id="nome"
            type="text"
            value={schoolValues.nome}
            onChange={(e) => handleChange(e, 'nome')}
          />
        </label>
        <label htmlFor="responsavel">
          Responsável:
          <input
            id="responsavel"
            type="text"
            value={schoolValues.responsavel}
            onChange={(e) => handleChange(e, 'responsavel')}
          />
        </label>
        <SchoolFormWrapper.Actions>
          <button
            type="submit"
            disabled={allowToSave}
          >
            Salvar
          </button>
          <button
            type="button"
            onClick={() => closeForm()}
          >
            Fechar
          </button>
        </SchoolFormWrapper.Actions>
        {}
      </form>
      {
        messageForSchoolSubmitted && <p>{messageForSchoolSubmitted}</p>
      }
    </SchoolFormWrapper>
  );
}

SchoolsForm.propTypes = {
  closeForm: PropTypes.func.isRequired,
};

export default SchoolsForm;
