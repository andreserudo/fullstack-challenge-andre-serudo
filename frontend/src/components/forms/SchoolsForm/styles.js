import styled from 'styled-components';

const SchoolFormWrapper = styled.div`
    width: 185px;
    display: flex;
    flex-direction: column;
    margin-top: 1.5rem;
    margin-bottom: 2rem;

    form {
        width: 100%;
        display: flex;
            flex-direction: column;
            align-items: center;

        label {
            display: flex;
            flex-direction: column;
            align-items: flex-start;
            
        }
    }
    
`;

SchoolFormWrapper.Actions = styled.div`
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-around;
    padding-top: 1.5rem;
`;

export default SchoolFormWrapper;
