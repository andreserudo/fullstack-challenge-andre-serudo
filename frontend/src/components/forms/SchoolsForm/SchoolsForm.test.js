import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ApplicationContext from '../../../context/ApplicationContext';

import SchoolsForm from './index';

describe('<SchoolsForm />', () => {
  it('should render according patterns', () => {
    const getAllSchools = jest.fn();
    const closeForm = jest.fn();

    render(
      <ApplicationContext.Provider value={{ getAllSchools }}>
        <SchoolsForm closeForm={closeForm} />
      </ApplicationContext.Provider>,

    );

    const nomeTextField = screen.getByText(/Nome da Escola:/i);
    const responsavelTextField = screen.getByText(/ResponsÃ¡vel:/i);
    const buttonToSave = screen.getByText(/Salvar/i);
    const buttonToClose = screen.getByText(/Fechar/i);
    expect(nomeTextField).toMatchSnapshot();
    expect(responsavelTextField).toMatchSnapshot();
    expect(buttonToSave).toMatchSnapshot();
    expect(buttonToClose).toMatchSnapshot();
  });

  it('should render Salvar button as disable', () => {
    const getAllSchools = jest.fn();
    const closeForm = jest.fn();
    render(
      <ApplicationContext.Provider value={{ getAllSchools }}>
        <SchoolsForm closeForm={closeForm} />
      </ApplicationContext.Provider>,
    );

    const buttonSalvar = screen.getByText(/Salvar/i);
    expect(buttonSalvar).toBeDisabled();
  });

  it('the button to Save should have be available only when all fields are filled', async () => {
    const getAllSchools = jest.fn();
    const closeForm = jest.fn();
    render(
      <ApplicationContext.Provider value={{ getAllSchools }}>
        <SchoolsForm closeForm={closeForm} />
      </ApplicationContext.Provider>,
    );

    const buttonSalvar = screen.getByText(/Salvar/i);
    await userEvent.type(screen.getByTestId('schoolNome'), 'Escola Estadual Augusto Nunes');
    await userEvent.type(screen.getByTestId('schoolResponsavel'), 'Maria do Carmo');
    expect(buttonSalvar).not.toBeDisabled();
  });
});
