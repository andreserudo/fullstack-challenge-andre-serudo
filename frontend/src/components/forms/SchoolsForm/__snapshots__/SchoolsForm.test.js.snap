// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SchoolsForm /> should render according patterns 1`] = `
<label
  for="nome"
>
  Nome da Escola:
  <input
    data-testid="schoolNome"
    id="nome"
    type="text"
    value=""
  />
</label>
`;

exports[`<SchoolsForm /> should render according patterns 2`] = `
<label
  for="responsavel"
>
  Respons√°vel:
  <input
    data-testid="schoolResponsavel"
    id="responsavel"
    type="text"
    value=""
  />
</label>
`;

exports[`<SchoolsForm /> should render according patterns 3`] = `
<button
  disabled=""
  type="submit"
>
  Salvar
</button>
`;

exports[`<SchoolsForm /> should render according patterns 4`] = `
<button
  type="button"
>
  Fechar
</button>
`;
